<h1 align=center style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
ارزش‌گذاری مشتریان
</font>
</h1>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
<font>
    در ادامه‌ی امور مربوط به بازاریابی، در این گام باید براساس ارزش مشتریان برای شرکت اقدام به گروه‌بندی آن‌ها کنید.
    برای این کار از یک روش معروف حوزه‌ی تحلیل رفتار مشتری با نام <a href="https://en.wikipedia.org/wiki/RFM_(market_research)"><i>RFM</i></a> استفاده خواهید کرد.
    در این روش برای هر مشتری سه معیار محاسبه می‌شود:
</font>
</p>


<div dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
<font>
    <ol>
        <li>
            <b>معیار <i>Recency</i> (ماندگاری):</b> تعداد روزی که از آخرین خرید مشتری می‌گذرد.
        </li>
        <li>
            <b>معیار <i>Frequency</i> (فرکانس):</b> تعداد دفعاتی که مشتری از ما خرید کرده است.
        </li>
        <li>
            <b>معیار <i>Monetary Value</i> (ارزش مالی):</b> مجموع درآمدی که شرکت از یک مشتری داشته است.
        </li>
    </ol>
</font>
</div>


<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font> 
این معیارها، نشان‌دهنده‌ی شاخص‌های مهمی در رفتار مشتریان هستند. به‌عنوان مثال، هرچه خرید جدیدتری توسط یک مشتری صورت گرفته باشد، احتمال آن‌که وی مجدد از ما خرید کند، افزایش می‌یابد. هرچه مشتری تعداد دفعات بیشتری خرید کند، به این معنی‌ست که مشتری رضایت بیشتری دارد و همچنین میزان درآمدی که یک مشتری برای شرکت تولید می‌کند، می‌تواند او را از سایر خریداران متمایز کند.
</font>
</p>


   
<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
کتابخانه‌ها
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
       ابتدا کتابخانه‌های مورد نیاز خود را فرا بخوانید.
</font>
</p>

In [126]:
import numpy as np
import pandas as pd

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    مجموعه‌داده
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    حال داده‌های پیش‌پردازش‌شده را که در گام اول پروژه در فایل <code>preprocessed_sales.csv</code> ذخیره کردید به شکل یک دیتافریم بخوانید.
</font>
</p>

In [127]:
df = pd.read_csv('preprocessed_sales.csv')
df

InvoiceNumber ProductCode                          ProductName  \
0              489434       85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS   
1              489434      79323P                   PINK CHERRY LIGHTS   
2              489434      79323W                  WHITE CHERRY LIGHTS   
3              489434       22041         RECORD FRAME 7" SINGLE SIZE    
4              489434       21232       STRAWBERRY CERAMIC TRINKET BOX   
...               ...         ...                                  ...   
400911         538171       22271                 FELTCRAFT DOLL ROSIE   
400912         538171       22750         FELTCRAFT PRINCESS LOLA DOLL   
400913         538171       22751       FELTCRAFT PRINCESS OLIVIA DOLL   
400914         538171       20970   PINK FLORAL FELTCRAFT SHOULDER BAG   
400915         538171       21931               JUMBO STORAGE BAG SUKI   

        Quantity          InvoiceDate  UnitPrice  CustomerId         Country  
0             12  2009-12-01 07:45:00       6.95     13085.0  United Kingdom  
1             12  2009-12-01 07:45:00       6.75     13085.0  United Kingdom  
2             12  2009-12-01 07:45:00       6.75     13085.0  United Kingdom  
3             48  2009-12-01 07:45:00       2.10     13085.0  United Kingdom  
4             24  2009-12-01 07:45:00       1.25     13085.0  United Kingdom  
...          ...                  ...        ...         ...             ...  
400911         2  2010-12-09 20:01:00       2.95     17530.0  United Kingdom  
400912         1  2010-12-09 20:01:00       3.75     17530.0  United Kingdom  
400913         1  2010-12-09 20:01:00       3.75     17530.0  United Kingdom  
400914         2  2010-12-09 20:01:00       3.75     17530.0  United Kingdom  
400915         2  2010-12-09 20:01:00       1.95     17530.0  United Kingdom  

[400916 rows x 8 columns]

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    معیار Recency
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    ابتدا نیاز است برای هر مشتری (<code>CustomerId</code>) معیار <i>Recency</i> را محاسبه کنید.
    برای این معیار باید تعداد روز گذشته از آخرین خرید کاربر نسبت به یک روز بعد از آخرین تاریخ موجود در مجموعه‌داده (ستون <code>InvoiceDate</code>) را محاسبه کنید. زیرا که فرض را بر آن گرفته‌ایم که شما در چنین روزی در حال تحلیل هستید.
</font>
</p>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    بنابراین ابتدا آخرین تاریخ موجود در مجموعه‌داده را پیدا کرده، و پس از جمع روز آن با <code>1</code>، حاصل را در متغیر <code>last_day</code> ذخیره کنید.
    مقدار این متغیر باید از نوع <code>Timestamp</code> باشد. به‌عنوان مثال اگر آخرین تاریخ معادل <code>Timestamp('2010-10-10 10:10:00')</code> بود، آن‌گاه حاصل شما باید به شکل <code>Timestamp('2010-10-11 10:10:00')</code> باشد.
</font>
</p>

In [128]:
df.InvoiceDate = pd.to_datetime(df.InvoiceDate)
last_day = df.InvoiceDate.max()
last_day = last_day + pd.Timedelta(days=1)
last_day

Timestamp('2010-12-10 20:01:00')

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    اکنون می‌توانید به ازای هر مشتری، تاریخ آخرین خرید وی را پیدا کرده و تعداد روز گذشته از <code>last_day</code> را محاسبه کنید.
    <br>
    حاصل را در یک دیتافریم به نام <code>df_customer_segments</code> و با دو ستون <code>CustomerId</code> (شناسه‌ی مشتری) و <code>Recency</code> (مقدار ماندگاری محاسبه‌شده) ذخیره کنید.
</font>
</p>

In [129]:
df_customer_segments = pd.DataFrame()
df_customer_segments['Recency'] = df.groupby('CustomerId').InvoiceDate.agg(lambda x: (last_day - x.max()).days)
df_customer_segments

Recency
CustomerId         
12346.0         165
12347.0           3
12348.0          74
12349.0          43
12351.0          11
...             ...
18283.0          18
18284.0          67
18285.0         296
18286.0         112
18287.0          18

[4312 rows x 1 columns]

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    معیار Frequency
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    اکنون باید محاسبه کنید هرکدام از مشتری‌ها تاکنون چند خرید <b>یکتا</b> داشته‌اند. حاصل را در ستونی به نام <code>Frequency</code> به دیتافریم <code>df_customer_segments</code> اضافه کنید.
</font>
</p>

In [130]:
df_customer_segments['Frequency'] = df.groupby('CustomerId')['InvoiceNumber'].nunique()
df_customer_segments

Recency  Frequency
CustomerId                    
12346.0         165         11
12347.0           3          2
12348.0          74          1
12349.0          43          3
12351.0          11          1
...             ...        ...
18283.0          18          6
18284.0          67          1
18285.0         296          1
18286.0         112          2
18287.0          18          4

[4312 rows x 2 columns]

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    معیار Monetary Value
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    برای این معیار باید محاسبه کنید مجموع خرید هر مشتری چقدر بوده است؟ توجه داشته باشید که برای محاسبه‌ی دریافتی هر ردیف باید قیمت واحد (<code>UnitPrice</code>) را در تعداد آن (<code>Quantity</code>) ضرب کنید و از حاصل آن برای محاسبه‌ی مجموع خرید هر مشتری استفاده کنید.
     حاصل را در ستونی به نام <code>MonetaryValue</code> به دیتافریم <code>df_customer_segments</code> اضافه کنید.
</font>
</p>

In [134]:
temp = df.copy()
temp['total'] = temp.Quantity * temp.UnitPrice
df_customer_segments['MonetaryValue'] = temp.groupby('CustomerId')['total'].sum()
df_customer_segments

Recency  Frequency  MonetaryValue  R_quartile  F_quartile  \
CustomerId                                                              
12346.0         165         11         372.86           4           4   
12347.0           3          2        1323.32           1           2   
12348.0          74          1         222.16           3           1   
12349.0          43          3        2671.14           2           3   
12351.0          11          1         300.93           1           1   
...             ...        ...            ...         ...         ...   
18283.0          18          6         619.37           1           4   
18284.0          67          1         461.68           3           1   
18285.0         296          1         427.00           4           1   
18286.0         112          2        1296.43           3           2   
18287.0          18          4        2345.71           1           3   

            M_quartile  RFM  
CustomerId                   
12346.0              2  442  
12347.0              3  123  
12348.0              1  311  
12349.0              4  234  
12351.0              1  111  
...                ...  ...  
18283.0              2  142  
18284.0              2  312  
18285.0              2  412  
18286.0              3  323  
18287.0              4  134  

[4312 rows x 7 columns]

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    گروه‌بندی هر معیار
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    اکنون نیاز است هرکدام از این سه معیار را براساس چارک اول،‌ دوم و سوم به شکل زیر به ۴ گروه مختلف تقسیم کنید
    (برای یادآوری مفهوم چارک،
    <a href="https://fa.wikipedia.org/wiki/%DA%86%D8%A7%D8%B1%DA%A9_(%D8%A2%D9%85%D8%A7%D8%B1)">این مطلب</a>
    را از ویکی‌پدیای فارسی بخوانید).
</font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>

| <b>شماره گروه</b> | <b>شرط</b> |
| :---: | :---: | 
| 1 | <code dir=ltr>value <= Q1</code> | 
| 2 | <code dir=ltr>Q1 < value <= Q2</code> | 
| 3 | <code dir=ltr>Q2 < value <= Q3</code> | 
| 4 | <code dir=ltr>Q3 < value</code> | 

</font>
</div>
</center>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    به ازای هر کدام از این سه معیار چنین کاری را انجام داده و نتایج را در سه ستون دیگر به نام‌های
    <code>F_quartile</code>، <code>R_quartile</code> و <code>M_quartile</code>
    ذخیره کنید.
    توجه کنید که مقادیر این ستون‌ها معادل شماره گروه یعنی اعداد <code>1</code> تا <code>4</code> خواهد بود.
</font>
</p>

In [135]:
def rfunc(x):
    if x <= rq1:
        return 1
    elif rq1 < x <= rq2:
        return 2
    elif rq2 < x <= rq3:
        return 3
    else:
        return 4

def ffunc(x):
    if x <= fq1:
        return 1
    elif fq1 < x <= fq2:
        return 2
    elif fq2 < x <= fq3:
        return 3
    else:
        return 4

def mfunc(x):
    if x <= mq1:
        return 1
    elif mq1 < x <= mq2:
        return 2
    elif mq2 < x <= mq3:
        return 3
    else:
        return 4

rq1 = np.quantile(df_customer_segments.Recency, 0.25)
rq2 = np.quantile(df_customer_segments.Recency, 0.5)
rq3 = np.quantile(df_customer_segments.Recency, 0.75)

fq1 = np.quantile(df_customer_segments.Frequency, 0.25)
fq2 = np.quantile(df_customer_segments.Frequency, 0.5)
fq3 = np.quantile(df_customer_segments.Frequency, 0.75)

mq1 = np.quantile(df_customer_segments.MonetaryValue, 0.25)
mq2 = np.quantile(df_customer_segments.MonetaryValue, 0.5)
mq3 = np.quantile(df_customer_segments.MonetaryValue, 0.75)

df_customer_segments['R_quartile'] = df_customer_segments['Recency'].apply(rfunc)
df_customer_segments['F_quartile'] = df_customer_segments['Frequency'].apply(ffunc)
df_customer_segments['M_quartile'] = df_customer_segments['MonetaryValue'].apply(mfunc)
df_customer_segments.head()

Recency  Frequency  MonetaryValue  R_quartile  F_quartile  \
CustomerId                                                              
12346.0         165         11         372.86           4           4   
12347.0           3          2        1323.32           1           2   
12348.0          74          1         222.16           3           1   
12349.0          43          3        2671.14           2           3   
12351.0          11          1         300.93           1           1   

            M_quartile  RFM  
CustomerId                   
12346.0              2  442  
12347.0              3  123  
12348.0              1  311  
12349.0              4  234  
12351.0              1  111

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    معیار RFM
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    اکنون می‌توانید با کنار هم قرار دادن مقادیر ستون‌های
    <code dir=ltr>[R_quartile, F_quartile, M_quartile]</code>
    مقدار نهایی <i>RFM</i> را محاسبه کرده و در ستونی به همین نام ذخیره کنید.
    به‌‌عنوان مثال اگر مقادیر این ستون‌ها برای یک مشتری برابر
    <code dir=ltr>[1, 2, 3]</code> بود،
    آن‌گاه مقدار نهایی <i>RFM</i> برابر با <code>123</code> است.
</font>
</p>

In [136]:
df_customer_segments['RFM'] = (df_customer_segments.R_quartile * 100 + df_customer_segments.F_quartile * 10 + df_customer_segments.M_quartile).astype('str')
df_customer_segments.head()

Recency  Frequency  MonetaryValue  R_quartile  F_quartile  \
CustomerId                                                              
12346.0         165         11         372.86           4           4   
12347.0           3          2        1323.32           1           2   
12348.0          74          1         222.16           3           1   
12349.0          43          3        2671.14           2           3   
12351.0          11          1         300.93           1           1   

            M_quartile  RFM  
CustomerId                   
12346.0              2  442  
12347.0              3  123  
12348.0              1  311  
12349.0              4  234  
12351.0              1  111

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    گروه‌بندی مشتریان
</font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    حال باید براساس مقادیر ستون <code>RFM</code> و طبق جدول زیر به گروه‌بندی مشتریان بپردازید و نام گروه آن‌ها را در ستونی به نام <code>Segment</code> ذخیره کنید.
    توجه داشته باشید که در این جدول، <code>X</code> به معنی هر مقداری است. به‌عنوان مثال <code>14X</code> به معنی <code>141</code>، <code>142</code>، <code>143</code> و <code>144</code> است.
</font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>

| <b>Segment</b> | <b>RFM</b> |
| :---: | :---: | 
| Best | <code dir=ltr>144</code> | 
| AlmostLost | <code dir=ltr>344</code> | 
| LostBigSpenders | <code dir=ltr>444</code> | 
| LostCheap | <code dir=ltr>441</code> | 
| Loyal | <code dir=ltr>X4X</code> | 
| BigSpenders | <code dir=ltr>XX4</code> | 

</font>
</div>
</center>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    <span style="color:purple"><b>نکته ۱:</b></span>
    در صورتی‌که یک مشتری در هیچ‌یک از گروه‌های جدول قرار نگرفت، گروه او را Normal اعلام کنید.
    <br>
    <span style="color:purple"><b>نکته ۲:</b></span>
    همچنین اگر یک RFM می‌تواند در دو گروه از جدول قرار بگیرد، اولویت با گروهی‌ست که در جدول، <b>بالاتر</b> آمده باشد.
</font>
</p>

In [138]:
def discretize_RFM(Value):
    Value = str(Value)
    if Value == '144':
        return 'Best'
    elif Value == '344':
        return 'AlmostLost'
    elif Value == '444':
        return 'LostBigSpenders'
    elif Value == '441':
        return 'LostCheap'
    elif Value[1:2] == '4':
        return'Loyal'
    elif Value[2:] == '4':
        return'BigSpenders'
    else:
        return 'Normal'

df_customer_segments['Segment'] = df_customer_segments['RFM'].apply(discretize_RFM)
df_customer_segments.reset_index(inplace=True)
df_customer_segments.head()

CustomerId  Recency  Frequency  MonetaryValue  R_quartile  F_quartile  \
0     12346.0      165         11         372.86           4           4   
1     12347.0        3          2        1323.32           1           2   
2     12348.0       74          1         222.16           3           1   
3     12349.0       43          3        2671.14           2           3   
4     12351.0       11          1         300.93           1           1   

   M_quartile  RFM      Segment  
0           2  442        Loyal  
1           3  123       Normal  
2           1  311       Normal  
3           4  234  BigSpenders  
4           1  111       Normal